sdk/core/azure-core/src/http/winhttp/win_http_transport.cpp

```diff
@@ -327,121 +329,335 @@ bool WinHttpTransport::VerifyCertificatesInChain(
+      DWORD statusInformationLength) noexcept
   {
-    if (dwInternetStatus == WINHTTP_CALLBACK_STATUS_SECURE_FAILURE)
+    // If we're called before our context has been set (on Open and Close callbacks), ignore the
+    // status callback.
+    if (dwContext == 0)
     {
-      Log::Write(Logger::Level::Error, "Security failure. :(");
+      return;
     }
-    // Silently ignore if there's any statuses we get that we can't handle
-    return;
-  }
 
-  // We will only set the Status callback if a root certificate has been set.
-  AZURE_ASSERT(!m_options.ExpectedTlsRootCertificates.empty());
+    try
+    {
+      WinHttpAction* httpAction = reinterpret_cast<WinHttpAction*>(dwContext);
+      httpAction->OnHttpStatusOperation(
+          hInternet, internetStatus, statusInformation, statusInformationLength);
+    }
+    catch (Azure::Core::RequestFailedException const& rfe)
+    {
+      // If an exception is thrown in the handler, log the error and terminate the connection.
+      Log::Write(
+          Logger::Level::Error,
+          "Request Failed Exception Thrown: " + std::string(rfe.what()) + rfe.Message);
+      WinHttpCloseHandle(hInternet);
+    }
+    catch (std::exception const& ex)
+    {
+      // If an exception is thrown in the handler, log the error and terminate the connection.
+      Log::Write(Logger::Level::Error, "Exception Thrown: " + std::string(ex.what()));
+    }
+  }
 
-  // Ask WinHTTP for the server certificate - this won't be valid outside a status callback.
-  wil::unique_cert_context serverCertificate;
+  /**
+   * @brief HTTP Callback to enable private certificate checks.
+   *
+   * This method is called by WinHTTP when a certificate is received. This method is called
+   * multiple times based on the state of the TLS connection.
+   *
+   * Special consideration for the WINHTTP_CALLBACK_STATUS_SENDING_REQUEST - this callback is
+   * called during the TLS connection - if a TLS roto certificate is configured, we verify that the
+   * certificate chain sent from the server contains the certificate the HTTP client was configured
+   * with. If it is, we accept the connection, if it is not, we abort the connection, closing the
+   * incoming request handle.
+   */
+  void WinHttpAction::OnHttpStatusOperation(
+      HINTERNET hInternet,
+      DWORD internetStatus,
+      LPVOID statusInformation,
+      DWORD statusInformationLength)
   {
-    DWORD bufferLength = sizeof(PCCERT_CONTEXT);
-    if (!WinHttpQueryOption(
-            hInternet,
-            WINHTTP_OPTION_SERVER_CERT_CONTEXT,
-            reinterpret_cast<void*>(serverCertificate.addressof()),
-            &bufferLength))
+    Log::Write(
+        Logger::Level::Informational,
+        "Status operation: " + std::to_string(internetStatus) + "("
+            + InternetStatusToString(internetStatus) + ")");
+    if (internetStatus == WINHTTP_CALLBACK_STATUS_SECURE_FAILURE)
+    {
+      Log::Write(Logger::Level::Error, "Security failure. :(");
```

Comment:
Where is this parenthesis after the colon in the log message closed?

Or is this a sad face emoji?

Reply:
That's a sad face emoji.